src/core/agent/services/signifyNotificationService.test.ts

```diff
@@ -669,4 +693,194 @@ describe("Signify notification service of agent", () => {
+    operationPendingGetAllMock.mockResolvedValueOnce([
+      {
+        type: "OperationPendingRecord",
+        id: "exchange.receivecredential.AOCUvGbpidkplC7gAoJOxLgXX1P2j4xlWMbzk3gM8JzA",
+        createdAt: new Date("2024-08-01T10:36:17.814Z"),
+        recordType: "exchange.receivecredential",
+        updatedAt: new Date("2024-08-01T10:36:17.814Z"),
+      },
+    ]);
+    const credentialIdMock = "credentialId";
+    signifyClient
+      .exchanges()
+      .get.mockResolvedValueOnce({
+        exn: {
+          r: ExchangeRoute.IpexAdmit,
+          p: "p",
+        },
+      })
+      .mockResolvedValueOnce({
+        exn: {
+          r: ExchangeRoute.IpexGrant,
+          e: {
+            acdc: {
+              d: credentialIdMock,
+            },
+          },
+        },
+      });
+    try {
+      await signifyNotificationService.onSignifyOperationStateChanged(callback);
+    } catch (error) {
+      expect((error as Error).message).toBe("Force Exit");
+    }
+    expect(Agent.agent.ipexCommunications.markAcdcComplete).toBeCalledWith(
+      credentialIdMock
+    );
+    expect(callback).toBeCalledTimes(3);
+    expect(operationPendingStorage.deleteById).toBeCalledTimes(4);
+  });
+
+  test("Should only delete the operation if the exchange route is not /ipex/admit", async () => {
+    Agent.agent.getKeriaOnlineStatus = jest.fn().mockReturnValue(true);
+    const callback = jest.fn();
+    const operationMock = {
+      metadata: {
+        said: "said",
+      },
+      done: true,
+      response: {
+        i: "id",
+        dt: new Date(),
+      },
+    };
+    operationsGetMock.mockResolvedValue(operationMock);
+    // We mock the setTimeout here so we can exit the while(true) loop
+    jest.spyOn(global, "setTimeout").mockImplementation(() => {
+      throw new Error("Force Exit");
+    });
+    operationPendingGetAllMock.mockResolvedValueOnce([
+      {
+        type: "OperationPendingRecord",
+        id: "exchange.receivecredential.AOCUvGbpidkplC7gAoJOxLgXX1P2j4xlWMbzk3gM8JzA",
+        createdAt: new Date("2024-08-01T10:36:17.814Z"),
+        recordType: "exchange.receivecredential",
+        updatedAt: new Date("2024-08-01T10:36:17.814Z"),
+      },
+    ]);
+    const credentialIdMock = "credentialId";
+    signifyClient.exchanges().get.mockResolvedValueOnce({
+      exn: {
+        r: ExchangeRoute.IpexGrant,
+        e: {
+          acdc: {
+            d: credentialIdMock,
+          },
+        },
+      },
+    });
+    try {
+      await signifyNotificationService.onSignifyOperationStateChanged(callback);
+    } catch (error) {
+      expect((error as Error).message).toBe("Force Exit");
+    }
+    expect(operationsGetMock).toBeCalledTimes(1);
+    expect(Agent.agent.ipexCommunications.markAcdcComplete).toBeCalledTimes(0);
+    expect(operationPendingStorage.deleteById).toBeCalledTimes(1);
+  });
+
+  test("Should call setTimeout  listening for pending operations if Keria is offline", async () => {
```

Comment:
because of the second setTimeout, this test would still pass if KERIA was online right? because it'd eventually hit the bottom setTimeout.

Reply:
Actually, I don't mock the `pendingOperations` here, so if it continues, it will throw this error `Cannot read properties of undefined (reading 'length')` instead, and the test will fail.

Reply:
Hmm ok, that is still a weak test because it could be refactored and become a false positive then. But I see now you changed it to `mockReturnValueOnce(true)` instead of false, why is that?

Reply:
I see it's failing now as you said. I think this would be a stronger test:
- mock `pendingOperations` properly
- set it to false as before
- expect the setTimeout error
- but also expect that `signifyClient.operations().get` is NOT called, even though there are `pendingOperations`

Reply:
oops, sorry, my bad, I just tried to test it with the `true` value again. I committed it mistakenly.